### PR TITLE
Added UDP listener

### LINKED ISSUE #26: UDP ingest protocol
We need to Implement UDP ingest protocol in nanotube, it is used on submission relays widely. Unfortunately, there is no easy way to measure the level of adoption.

Originally submitted by @azhiltsov

### PR DESCRIPTION
**Pull request by grzkv**
This adds a listener on UDP port in addition to TCP.

This fixes #26

### CODE REVIEW

File: cmd/nanotube/listen.go [hunk 1]
```diff
@@ -86,6 +109,11 @@ func readFromConnection(wg *sync.WaitGroup, conn net.Conn, queue chan string, st
 			zap.String("sender", conn.RemoteAddr().String()))
 	}
 
+	scanForRecords(conn, queue, stop, ms)
+}
+
+func scanForRecords(r io.Reader, queue chan string, stop <-chan struct{}, ms *metrics.Prom) {
+	sc := bufio.NewScanner(r)
```
Comment (avereha) [inline cmd/nanotube/listen.go]: UDP is not playing well with bufio.
The reason is that one read() from  UDP will return back one datagram.
If we use bufio, and buffer size(at bufio level)  is < length of datagram, the unread part is discarded&lost.
See also: https://linux.die.net/man/2/recv
In this case, I would not use any buffering from UDP, we could just call strings.split() by \n for data coming from UDP and call Read() by hand.

File: cmd/nanotube/listen.go [hunk 2]
```diff
@@ -16,20 +17,23 @@ import (
 
 // Listen listens for incoming metric data
 func Listen(cfg *conf.Main, stop <-chan struct{}, lg *zap.Logger, ms *metrics.Prom) (chan string, error) {
+	queue := make(chan string, cfg.MainQueueSize)
+
 	l, err := net.Listen("tcp", fmt.Sprintf(":%d", cfg.ListeningPort))
```
[the remark this answers is not in the document]
Comment (grzkv) [reply cmd/nanotube/listen.go]: This makes sense.

File: cmd/nanotube/listen.go (hunk 1 above)
Comment (grzkv) [reply cmd/nanotube/listen.go]: I am not sure we are discussing the same problem but there is this thing
https://github.com/golang/go/blob/531b6d3126bb810dcdfcce14983cea32236c322d/src/bufio/scan.go#L188
the scanner buffer is sized to hold one token (one line in our case). Max is 64k.

Comment (grzkv) [reply cmd/nanotube/listen.go]: Do you think it will still cut the input buffer and we will lose the data?

Comment (avereha) [reply cmd/nanotube/listen.go]: You should be able to reproduce this by connecting and sending one datagram > startBufSize.

File: cmd/nanotube/listen.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (grzkv) [reply cmd/nanotube/listen.go]: implemented

File: cmd/nanotube/listen.go (hunk 1 above)
Comment (grzkv) [reply cmd/nanotube/listen.go]: Yes, you are right. The scanner reads the startBufSize sized buffer. The buffer extension is only used if no token delimiter is encountered. It is not used to accommodate large one-time inputs.

File: cmd/nanotube/listen.go [hunk 3]
```diff
@@ -16,20 +17,25 @@ import (
...
+		if err != nil {
+			return nil, errors.Wrap(err,
+				"error while opening TCP port for listening")
+		}
+		go acceptAndListenTCP(l, queue, stop, cfg, ms, lg)
+	}
+
+	if cfg.EnableUDP {
+		go listenUDP(cfg, queue, stop, lg, ms)
+	}
```
Comment (avereha) [inline cmd/nanotube/listen.go]: Check that at least one is true here.

Comment (grzkv) [reply cmd/nanotube/listen.go]: It is checked during config parsing.

**Comment by grzkv** [comment]
We don't have to handle residue if there are not following records. If we stop and the residue is non-empty, we have no other choice but to ignore it because it represents an incomplete record.

**Comment by avereha** [comment]
Are there actual clients that send data this way(one line splitted over two datagrams)?
In not, we are only increasing complexity without getting anything out of it.

**Comment by grzkv** [comment]
> Are there actual clients that send data this way(one line split over two datagrams)? If there are, this is most likely not correct.
> In not, we are only increasing complexity without getting anything out of it.
> This is also incorrect for multiple reasons:
> a) UDP has no ordering guarantee.
> b) 2 sequent datagrams can come from two different clients. Considering one continuation of the other will not be correct.

I think I got you wrong. I will rework it.

**Comment by grzkv** [comment]
@avereha Let's do it again! Plz review.

**Comment by grzkv** [comment]
I also tweaked the metrics a bit.

File: cmd/nanotube/listen.go [hunk 4]
```diff
@@ -74,12 +73,14 @@ loop:
 }
 
 func listenUDP(cfg *conf.Main, queue chan string, stop <-chan struct{}, lg *zap.Logger, ms *metrics.Prom) {
+	// TODO (grzkv) What should we do if connection drops?
```
Comment (avereha) [inline cmd/nanotube/listen.go]: Crash with error message.
Common use case would be when there is another instance running&listening on the same port. We don't want to have the new instance hanging around if it fails to Listen(). 
Better to crash and let systemd/other system restart.

File: cmd/nanotube/listen.go [hunk 5]
```diff
@@ -92,18 +93,16 @@ func listenUDP(cfg *conf.Main, queue chan string, stop <-chan struct{}, lg *zap.
...
 	}()
 
 	buf := make([]byte, 64*1024)
 
-	residue := ""
 	in := make(chan struct{})
 	errCh := make(chan error)
 loop:
 	for {
 		go func() {
```
Comment (avereha) [inline cmd/nanotube/listen.go]: You can benchmark this, I guess that having one new goroutine/datagram would be pretty expensive - and not needed, one single routine should be able to just do the Reads(). 
Not sure about the split() that we do later.

Comment (grzkv) [reply cmd/nanotube/listen.go]: Yes, that is not good thx.

File: cmd/nanotube/listen.go (hunk 4 above)
Comment (grzkv) [reply cmd/nanotube/listen.go]: This todo is more about "what to do if the connection drops in the middle of transmission".

Comment (avereha) [reply cmd/nanotube/listen.go]: Ok. This comment belongs two lines below, will move it.

File: cmd/nanotube/listen.go [hunk 6]
```diff
@@ -66,17 +72,86 @@ loop:
...
-func readFromConnection(wg *sync.WaitGroup, conn net.Conn, queue chan string, stop <-chan struct{}, cfg *conf.Main, ms *metrics.Prom, lg *zap.Logger) {
+func listenUDP(cfg *conf.Main, queue chan string, stop <-chan struct{}, lg *zap.Logger, ms *metrics.Prom) {
+	// TODO (grzkv) What should we do if connection drops?
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{
+		IP:   nil,
+		Port: int(cfg.ListeningPort),
+		Zone: ""})
+	if err != nil {
+		lg.Error("error while opening UDP port for listening", zap.Error(err))
+		return
```
Comment (avereha) [inline cmd/nanotube/listen.go]: Crash with error message/return this error back.
Common use case would be when there is another instance running&listening on the same port. We don't want to have the new instance hanging around if it fails to Listen().
Better to crash and let systemd/other system restart.

File: cmd/nanotube/listen.go [hunk 7]
```diff
@@ -66,17 +72,86 @@ loop:
...
+	}
+
+	err = conn.SetReadBuffer(64 * 1024)
```
Comment (avereha) [inline cmd/nanotube/listen.go]: Make this configurable.

**Pull request merged**